src/storage/src/compaction/scheduler.rs

```diff
@@ -0,0 +1,449 @@
+}
+
+pub trait CompactionRequest: Send + Sync + Default + 'static {
+    fn table_id(&self) -> TableId;
+}
+
+#[derive(Debug)]
+pub struct CompactionSchedulerConfig {
+    max_inflight_task: usize,
+}
+
+impl Default for CompactionSchedulerConfig {
+    fn default() -> Self {
+        Self {
+            max_inflight_task: 16,
+        }
+    }
+}
+
+/// CompactionScheduler defines a set of API to schedule compaction tasks.
+#[async_trait]
+pub trait CompactionScheduler<R> {
+    /// Schedules a compaction request.
+    /// Returns true if request is scheduled. Returns false if task queue already
+    /// contains the request with same table id.
+    async fn schedule(&self, request: R) -> Result<bool>;
+
+    /// Stops compaction scheduler.
+    async fn stop(&self) -> Result<()>;
+}
+
+/// Compaction task scheduler based on local state.
+#[allow(unused)]
+pub struct LocalCompactionScheduler<R: CompactionRequest> {
+    request_queue: Arc<RwLock<DedupDeque<TableId, R>>>,
+    cancel_token: CancellationToken,
+    task_notifier: Arc<Notify>,
+    join_handle: Mutex<Option<JoinHandle<()>>>,
+}
+
+#[async_trait]
+impl<R> CompactionScheduler<R> for LocalCompactionScheduler<R>
+where
+    R: CompactionRequest + Send + Sync,
+{
+    async fn schedule(&self, request: R) -> Result<bool> {
+        debug!(
+            "Schedule request: {}, queue size: {}",
+            request.table_id(),
+            self.remaining_requests().await
+        );
+        let mut queue = self.request_queue.write().await;
+        let res = queue.push_back(request.table_id(), request);
+        self.task_notifier.notify_one();
+        Ok(res)
+    }
+
+    async fn stop(&self) -> Result<()> {
+        self.cancel_token.cancel();
+        // safety: LocalCompactionScheduler is guaranteed to have a present join_handle field
+        let handle = { self.join_handle.lock().unwrap().take().unwrap() };
+        handle.await.context(StopCompactionSchedulerSnafu)?;
+        Ok(())
+    }
+}
+
+#[allow(unused)]
+impl<R> LocalCompactionScheduler<R>
+where
+    R: CompactionRequest,
+{
+    pub fn new<P, T>(config: CompactionSchedulerConfig, picker: P) -> Self
+    where
+        T: CompactionTask,
+        P: Picker<R, T> + Send + Sync,
+    {
+        let request_queue: Arc<RwLock<DedupDeque<TableId, R>>> =
+            Arc::new(RwLock::new(DedupDeque::default()));
+        let cancel_token = CancellationToken::new();
+        let task_notifier = Arc::new(Notify::new());
+
+        let handler = CompactionHandler {
+            task_notifier: task_notifier.clone(),
+            req_queue: request_queue.clone(),
+            cancel_token: cancel_token.child_token(),
+            limiter: Arc::new(CascadeRateLimiter::new(vec![Box::new(
+                MaxInflightTaskLimiter::new(config.max_inflight_task),
+            )])),
+            picker,
+            _phantom_data: PhantomData::<T>::default(),
+        };
+        let join_handle = common_runtime::spawn_bg(async move {
```

Comment:
Is it possible to reuse the background job facility in `storage::background`? e.g. Replace the CompactionTask trait with the Job trait; Submit the task to the JobPool (we might need to refactor the JobPool)

Reply:
> Is it possible to reuse the background job facility in `storage::background`? e.g. Replace the CompactionTask trait with the Job trait; Submit the task to the JobPool (we might need to refactor the JobPool)

Looks like the `storage::background` crate lacks some components like FIFO scheduling and rate limit. Maybe I should adapt the scheduler to `JobPool` trait and promote these components to a separate crate like `common/scheduler`.